File: packages/vaultsync/vaultsync/authenticated_state_view.py
"""Bounded client evidence comparison; not a freshness or consensus protocol."""

from __future__ import annotations

import base64
import hashlib
import threading
from contextlib import contextmanager
from pathlib import Path

from cryptography.exceptions import InvalidSignature
from cryptography.hazmat.primitives.asymmetric.ed25519 import Ed25519PublicKey
from cryptography.hazmat.primitives.serialization import Encoding, PublicFormat

from .sync_queue import (
    _ROOT_SIGNATURE_DOMAIN,
    SignedStateCommitment,
    _canonical_base64,
    _commitment_hex,
    _commitment_sequence,
    _EncryptedQueueFile,
    _identifier,
    _state_digest,
)

FIELDS = (
    "account_id",
    "vault_id",
    "sequence",
    "previous_root",
    "collection_root",
    "registry_root",
    "previous_registry_root",
    "key_epoch",
)
FORMAT = "atlasvault-authenticated-state-view"
ZERO = "0" * 64
EMPTY_REGISTRY = hashlib.sha256(b"atlasvault-registry-root-v1\n").hexdigest()
LIMIT = 256


class StateViewError(ValueError):
    """Stable secret-free failure code."""


def _reject(code="ATLAS_STATE_VIEW_REJECTED"):
    raise StateViewError(code) from None


@contextmanager
def _boundary():
    try:
        yield
    except StateViewError:
        raise
    except (ValueError, TypeError, KeyError, OSError, InvalidSignature):
        raise StateViewError("ATLAS_STATE_VIEW_REJECTED") from None


def registry_root(entries):
    with _boundary():
        if not isinstance(entries, list) or not 1 <= len(entries) <= LIMIT:
            _reject("ATLAS_REGISTRY_SUBSTITUTION")
        checked = {}
        for entry in entries:
            if not isinstance(entry, dict) or set(entry) != {"device_id", "descriptor_sha256"}:
                _reject("ATLAS_REGISTRY_SUBSTITUTION")
            device = _commitment_hex(entry["device_id"])
            descriptor = _commitment_hex(entry["descriptor_sha256"])
            if device in checked:
                _reject("ATLAS_REGISTRY_SUBSTITUTION")
            checked[device] = descriptor
        transcript = "atlasvault-registry-root-v1\n" + "".join(
            f"{k}:{checked[k]}\n" for k in sorted(checked)
        )
        return hashlib.sha256(transcript.encode("ascii")).hexdigest()


def _root(unsigned):
    if set(unsigned) != {"format", "version", *FIELDS}:
        _reject()
    if (
        unsigned["format"] != FORMAT
        or type(unsigned["version"]) is not int
        or unsigned["version"] != 2
    ):
        _reject()
    for field in FIELDS:
        if field in ("account_id", "vault_id"):
            _identifier(unsigned[field])
        elif field in ("sequence", "key_epoch"):
            _commitment_sequence(unsigned[field])
        else:
            _commitment_hex(unsigned[field])
    transcript = "atlasvault-authenticated-state-view-v2\n" + "".join(
        str(unsigned[k]) + "\n" for k in FIELDS
    )
    return hashlib.sha256(transcript.encode("ascii")).hexdigest()


def _message(root):
    return b"atlasvault-state-view-signature-v2\0" + bytes.fromhex(root)


def _verified(view, public):
    if not isinstance(view, dict) or set(view) != {
        "format",
        "version",
        *FIELDS,
        "root",
        "signature_b64",
    }:
        _reject()
    value = dict(view)
    unsigned = {k: v for k, v in value.items() if k not in ("root", "signature_b64")}
    if _commitment_hex(value["root"]) != _root(unsigned):
        _reject()
    if not isinstance(value["signature_b64"], str) or len(value["signature_b64"]) != 88:
        _reject()
    signature = base64.b64decode(_canonical_base64(value["signature_b64"], exact_bytes=64))
    Ed25519PublicKey.from_public_bytes(public).verify(signature, _message(value["root"]))
    return value


def sign_state_view(unsigned, signing_key):
    with _boundary():
        root = _root(unsigned)
        value = dict(
            unsigned,
            root=root,
            signature_b64=base64.b64encode(signing_key.sign(_message(root))).decode("ascii"),
        )
        return _verified(
            value, signing_key.public_key().public_bytes(Encoding.Raw, PublicFormat.Raw)
        )


class AuthenticatedHistory:
    def __init__(
        self,
        path: Path,
        *,
        encryption_key: bytes,
        account_id: str,
        vault_id: str,
        collection_id: str,
        key_epoch: int,
        trusted_signer: bytes,
    ):
        with _boundary():
            if not isinstance(trusted_signer, bytes) or len(trusted_signer) != 32:
                _reject()
            self._public = trusted_signer
            self._context = {
                "account_id": _identifier(account_id),
                "vault_id": _identifier(vault_id),
                "collection_id": _identifier(collection_id),
                "key_epoch": _commitment_sequence(key_epoch),
                "signing_public_b64": base64.b64encode(trusted_signer).decode("ascii"),
            }
            self._store = _EncryptedQueueFile(
                Path(path), encryption_key=encryption_key, kind="authenticated-state-history-v2"
            )
            self._lock = threading.Lock()

    def _chain(self, views):
        if not isinstance(views, list) or len(views) > LIMIT:
            _reject("ATLAS_HISTORY_LIMIT")
        checked = []
        previous, registry = ZERO, EMPTY_REGISTRY
        for i, raw in enumerate(views):
            view = _verified(raw, self._public)
            if any(view[k] != self._context[k] for k in ("account_id", "vault_id", "key_epoch")):
                _reject()
            if (
                view["sequence"] != i + 1
                or view["previous_root"] != previous
                or view["previous_registry_root"] != registry
            ):
                _reject()
            checked.append(view)
            previous, registry = view["root"], view["registry_root"]
        return checked

    def _load(self):
        value = self._store.read({})
        if (
            set(value) != {"context", "views", "blocked"}
            or value["context"] != self._context
            or type(value["blocked"]) is not bool
        ):
            _reject()
        return self._chain(value["views"]), value["blocked"]

    def _save(self, views, blocked=False):
        self._store.write({"context": self._context, "views": views, "blocked": blocked})

    def initialize(self):
        with self._lock, _boundary():
            if self._store.path.exists():
                _reject()
            self._save([])

    def export_evidence(self):
        with self._lock, _boundary():
            return self._load()[0]

    def _fork(self, views):
        self._save(views, blocked=True)
        _reject("ATLAS_STATE_EQUIVOCATION")

    def compare_evidence(self, peer):
        with self._lock, _boundary():
            local, blocked = self._load()
            if blocked:
                _reject("ATLAS_STATE_EQUIVOCATION")
            peer = self._chain(peer)
            if not local or not peer:
                _reject("ATLAS_CHECKPOINT_REQUIRED")
            for left, right in zip(local, peer):
                if left["root"] != right["root"]:
                    self._fork(local)
            return min(len(local), len(peer))

    def observe(self, raw_view, registry, raw_collection, opaque_state):
        with self._lock, _boundary():
            views, blocked = self._load()
            if blocked:
                _reject("ATLAS_STATE_EQUIVOCATION")
            view = _verified(raw_view, self._public)
            if any(view[k] != self._context[k] for k in ("account_id", "vault_id", "key_epoch")):
                _reject()
            if registry_root(registry) != view["registry_root"]:
                _reject("ATLAS_REGISTRY_SUBSTITUTION")
            collection = SignedStateCommitment.from_dict(raw_collection)
            if (
                collection.collection_id != self._context["collection_id"]
                or collection.sequence != view["sequence"]
                or collection.root != view["collection_root"]
                or collection.state_sha256 != _state_digest(opaque_state)
            ):
                _reject()
            Ed25519PublicKey.from_public_bytes(self._public).verify(
                base64.b64decode(collection.signature_b64),
                _ROOT_SIGNATURE_DOMAIN + bytes.fromhex(collection.root),
            )
            sequence = view["sequence"]
            if sequence <= len(views) and view["root"] != views[sequence - 1]["root"]:
                self._fork(views)
            if sequence < len(views):
                _reject("ATLAS_ROLLBACK_REJECTED")
            if sequence == len(views):
                return False
            if len(views) == LIMIT:
                _reject("ATLAS_HISTORY_LIMIT")
            previous = views[-1] if views else None
            if (
                sequence != len(views) + 1
                or view["previous_root"] != (previous["root"] if previous else ZERO)
                or view["previous_registry_root"]
                != (previous["registry_root"] if previous else EMPTY_REGISTRY)
                or collection.previous_root != (previous["collection_root"] if previous else ZERO)
            ):
                _reject()
            self._save([*views, view])
            return True

File: apps/atlas_flutter/lib/src/atlas_vault/authenticated_state_view.dart
part of 'sync_queue.dart';

const _viewFields = [
  'account_id',
  'vault_id',
  'sequence',
  'previous_root',
  'collection_root',
  'registry_root',
  'previous_registry_root',
  'key_epoch',
];
const _viewFormat = 'atlasvault-authenticated-state-view';
const _viewLimit = 256;
final _emptyRegistryRoot = _sha256Hex(
  ascii.encode('atlasvault-registry-root-v1\n'),
);

final class AtlasVaultStateViewException implements Exception {
  const AtlasVaultStateViewException(this.code);
  final String code;
  @override
  String toString() => code;
}

Never _viewFail([String code = 'ATLAS_STATE_VIEW_REJECTED']) =>
    throw AtlasVaultStateViewException(code);

final class AtlasVaultAuthenticatedStateView {
  static String registryRoot(List<Map<String, Object?>> entries) {
    try {
      if (entries.isEmpty || entries.length > _viewLimit) {
        _viewFail('ATLAS_REGISTRY_SUBSTITUTION');
      }
      final checked = <String, String>{};
      for (final e in entries) {
        _exact(e, {'device_id', 'descriptor_sha256'});
        final id = _commitmentHex(e['device_id']),
            descriptor = _commitmentHex(e['descriptor_sha256']);
        if (checked.containsKey(id)) _viewFail('ATLAS_REGISTRY_SUBSTITUTION');
        checked[id] = descriptor;
      }
      final ids = checked.keys.toList()..sort();
      return _sha256Hex(
        ascii.encode(
          'atlasvault-registry-root-v1\n${ids.map((id) => '$id:${checked[id]}\n').join()}',
        ),
      );
    } on AtlasVaultStateViewException {
      rethrow;
    } catch (_) {
      _viewFail();
    }
  }

  static String _root(Map<String, Object?> unsigned) {
    _exact(unsigned, {'format', 'version', ..._viewFields});
    if (unsigned['format'] != _viewFormat ||
        unsigned['version'] is! int ||
        unsigned['version'] != 2) {
      _viewFail();
    }
    for (final f in _viewFields) {
      if (f == 'account_id' || f == 'vault_id') {
        _commitmentIdentifier(unsigned[f]);
      } else if (f == 'sequence' || f == 'key_epoch') {
        _commitmentSequence(unsigned[f]);
      } else {
        _commitmentHex(unsigned[f]);
      }
    }
    return _sha256Hex(
      ascii.encode(
        'atlasvault-authenticated-state-view-v2\n${_viewFields.map((f) => '${unsigned[f]}\n').join()}',
      ),
    );
  }

  static Uint8List _message(String root) => Uint8List.fromList([
    ...ascii.encode('atlasvault-state-view-signature-v2\x00'),
    for (var i = 0; i < 64; i += 2)
      int.parse(root.substring(i, i + 2), radix: 16),
  ]);

  static Future<Map<String, Object?>> _verify(
    Map<String, Object?> view,
    Uint8List public,
  ) async {
    _exact(view, {
      'format',
      'version',
      ..._viewFields,
      'root',
      'signature_b64',
    });
    final value = Map<String, Object?>.of(view);
    final unsigned = Map<String, Object?>.of(value)
      ..remove('root')
      ..remove('signature_b64');
    final root = _commitmentHex(value['root']);
    if (root != _root(unsigned) ||
        value['signature_b64'] is! String ||
        (value['signature_b64']! as String).length != 88) {
      _viewFail();
    }
    final valid = await Ed25519().verify(
      _message(root),
      signature: Signature(
        _base64(value['signature_b64'], exactLength: 64),
        publicKey: SimplePublicKey(public, type: KeyPairType.ed25519),
      ),
    );
    if (!valid) _viewFail();
    return value;
  }

  static Future<Map<String, Object?>> sign(
    Map<String, Object?> unsigned,
    SimpleKeyPair signingKey,
  ) async {
    try {
      final fields = Map<String, Object?>.of(unsigned), root = _root(unsigned);
      final signature = await Ed25519().sign(
        _message(root),
        keyPair: signingKey,
      );
      return await _verify({
        ...fields,
        'root': root,
        'signature_b64': base64Encode(signature.bytes),
      }, Uint8List.fromList((await signingKey.extractPublicKey()).bytes));
    } on AtlasVaultStateViewException {
      rethrow;
    } catch (_) {
      _viewFail();
    }
  }
}

/// Explicit, bounded evidence exchange. One owner per persisted history path.
final class AtlasVaultAuthenticatedHistory {
  AtlasVaultAuthenticatedHistory({
    required File file,
    required Uint8List encryptionKey,
    required String accountId,
    required String vaultId,
    required String collectionId,
    required int keyEpoch,
    required Uint8List trustedSigner,
  }) : _store = _EncryptedQueueFile(
         file,
         encryptionKey,
         kind: 'authenticated-state-history-v2',
       ),
       _public = Uint8List.fromList(trustedSigner),
       _context = {
         'account_id': _commitmentIdentifier(accountId),
         'vault_id': _commitmentIdentifier(vaultId),
         'collection_id': _commitmentIdentifier(collectionId),
         'key_epoch': _commitmentSequence(keyEpoch),
         'signing_public_b64': base64Encode(trustedSigner),
       } {
    if (trustedSigner.length != 32) _viewFail();
  }
  final _EncryptedQueueFile _store;
  final Uint8List _public;
  final Map<String, Object?> _context;
  bool _busy = false;

  Future<T> _exclusive<T>(Future<T> Function() work) async {
    if (_busy) _viewFail();
    _busy = true;
    try {
      return await work();
    } on AtlasVaultStateViewException {
      rethrow;
    } catch (_) {
      _viewFail();
    } finally {
      _busy = false;
    }
  }

  Future<List<Map<String, Object?>>> _chain(
    List<Map<String, Object?>> views,
  ) async {
    if (views.length > _viewLimit) _viewFail('ATLAS_HISTORY_LIMIT');
    final checked = <Map<String, Object?>>[];
    var previous = _zeroRoot, registry = _emptyRegistryRoot;
    for (var i = 0; i < views.length; i++) {
      final v = await AtlasVaultAuthenticatedStateView._verify(
        views[i],
        _public,
      );
      if ([
        'account_id',
        'vault_id',
        'key_epoch',
      ].any((f) => v[f] != _context[f])) {
        _viewFail();
      }
      if (v['sequence'] != i + 1 ||
          v['previous_root'] != previous ||
          v['previous_registry_root'] != registry) {
        _viewFail();
      }
      checked.add(v);
      previous = v['root']! as String;
      registry = v['registry_root']! as String;
    }
    return checked;
  }

  Future<({List<Map<String, Object?>> views, bool blocked})> _load() async {
    final s = await _store.read({});
    _exact(s, {'context', 'views', 'blocked'});
    final context = _object(s['context']);
    _exact(context, _context.keys.toSet());
    if (_context.entries.any((e) => context[e.key] != e.value) ||
        s['blocked'] is! bool ||
        s['views'] is! List) {
      _viewFail();
    }
    final raw = s['views']! as List;
    if (raw.length > _viewLimit) _viewFail('ATLAS_HISTORY_LIMIT');
    return (
      views: await _chain(raw.map(_object).toList()),
      blocked: s['blocked']! as bool,
    );
  }

  Future<void> _save(
    List<Map<String, Object?>> views, {
    bool blocked = false,
  }) => _store.write({'context': _context, 'views': views, 'blocked': blocked});
  Future<void> initialize() => _exclusive(() async {
    if ((await _store.read({})).isNotEmpty || await _store.file.exists()) {
      _viewFail();
    }
    await _save([]);
  });
  Future<List<Map<String, Object?>>> exportEvidence() =>
      _exclusive(() async => (await _load()).views);
  Future<Never> _fork(List<Map<String, Object?>> views) async {
    await _save(views, blocked: true);
    _viewFail('ATLAS_STATE_EQUIVOCATION');
  }

  Future<int> compareEvidence(List<Map<String, Object?>> peer) async {
    if (peer.length > _viewLimit) _viewFail('ATLAS_HISTORY_LIMIT');
    final copied = peer.map((v) => Map<String, Object?>.of(v)).toList();
    return _exclusive(() async {
      final local = await _load();
      if (local.blocked) _viewFail('ATLAS_STATE_EQUIVOCATION');
      final checked = await _chain(copied);
      if (local.views.isEmpty || checked.isEmpty) {
        _viewFail('ATLAS_CHECKPOINT_REQUIRED');
      }
      final common = local.views.length < checked.length
          ? local.views.length
          : checked.length;
      for (var i = 0; i < common; i++) {
        if (local.views[i]['root'] != checked[i]['root']) {
          return _fork(local.views);
        }
      }
      return common;
    });
  }

  Future<bool> observe(
    Map<String, Object?> rawView,
    List<Map<String, Object?>> registry,
    Map<String, Object?> rawCollection,
    Uint8List opaqueState,
  ) async {
    final view = Map<String, Object?>.of(rawView),
        collection = Map<String, Object?>.of(rawCollection);
    final registryDigest = AtlasVaultAuthenticatedStateView.registryRoot(
      registry,
    );
    if (opaqueState.length < 16 || opaqueState.length > _maximumQueueBytes) {
      _viewFail();
    }
    final stateDigest = _commitmentDigest(opaqueState);
    return _exclusive(() async {
      final local = await _load();
      if (local.blocked) _viewFail('ATLAS_STATE_EQUIVOCATION');
      final v = await AtlasVaultAuthenticatedStateView._verify(view, _public);
      if ([
        'account_id',
        'vault_id',
        'key_epoch',
      ].any((f) => v[f] != _context[f])) {
        _viewFail();
      }
      if (v['registry_root'] != registryDigest) {
        _viewFail('ATLAS_REGISTRY_SUBSTITUTION');
      }
      final c = AtlasVaultSignedStateCommitment.fromJson(collection);
      if (c.collectionId != _context['collection_id'] ||
          c.sequence != v['sequence'] ||
          c.root != v['collection_root'] ||
          c._value['state_sha256'] != stateDigest) {
        _viewFail();
      }
      if (!await Ed25519().verify(
        _rootSignatureMessage(c.root),
        signature: Signature(
          _base64(c._value['signature_b64'], exactLength: 64),
          publicKey: SimplePublicKey(_public, type: KeyPairType.ed25519),
        ),
      )) {
        _viewFail();
      }
      final sequence = v['sequence']! as int, views = local.views;
      if (sequence <= views.length &&
          v['root'] != views[sequence - 1]['root']) {
        return _fork(views);
      }
      if (sequence < views.length) _viewFail('ATLAS_ROLLBACK_REJECTED');
      if (sequence == views.length) return false;
      if (views.length == _viewLimit) _viewFail('ATLAS_HISTORY_LIMIT');
      final previous = views.isEmpty ? null : views.last;
      if (sequence != views.length + 1 ||
          v['previous_root'] != (previous?['root'] ?? _zeroRoot) ||
          v['previous_registry_root'] !=
              (previous?['registry_root'] ?? _emptyRegistryRoot) ||
          c.previousRoot != (previous?['collection_root'] ?? _zeroRoot)) {
        _viewFail();
      }
      await _save([...views, v]);
      return true;
    });
  }
}
